### PR TITLE
Add model store managers

### PR DESCRIPTION
**Pull request by caisq**
- Add interface
- Add implementation for local storage:
  tf.io.browserLocalStorageManager()
- Add implementation for IndexedDB:
  tf.io.browserIndexedDBManager()
- Add public API for
  * Listing models: e.g., `tf.io.listModels();`
  * Copying model: e.g., `tf.io.copyModel('indexeddb://foo', 'localstorage://bar');`
  * Moving model: e.g., `tf.io.moveModel('indexeddb://foo', 'localstorage://bar');`
  * Removing model: e.g., `tf.io.removeModel('indexeddb://foo');`

Each of the two implementation supports:
- Listing all models: their paths and some meta data
  (ModelArtifactsInfo), such as topology type,
  byte sizes of topology and weights, etc.
- Deleting a model by URL.
- Copying a model by URL.

Change the IndexedDB to two tables (object stores) for
efficient look up of ModelArtifactsInfo:
- One for ModelArtifactsInfo (faster access)
- One for the actual artifacts (slower access)

Also, change "KerasJSON" to "JSON"

Towards: https://github.com/tensorflow/tfjs/issues/13

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038)

### CODE REVIEW

File: src/io/types.ts [hunk 1]
```diff
@@ -208,3 +208,47 @@ export interface IOHandler {
...
+   *   model artifacts info. Model artifacts info include type of the model's
+   *   topology, byte sizes of the topology, weights, etc.
+   */
+  listModels(): Promise<{[path: string]: ModelArtifactsInfo}>;
+
+  /**
+   * Delete a model specified by `path`.
+   *
+   * @param path
+   * @returns ModelArtifactsInfo of the deleted model (if and only if deletion
```
Comment (bileschi) [inline src/io/types.ts]: Why does deleting return the ModelArtifactsInfo?    Is this a common pattern for the deletion of assets in JS?  Are we creating a difficult-to-fix problem if some storage medium can sense presence & delete much faster than access the ModelArtifactsInfo?

File: src/io/types.ts [hunk 2]
```diff
@@ -208,3 +208,47 @@ export interface IOHandler {
...
+   * @throws Error if deletion fails, e.g., if no model exists at `path`.
+   */
+  deleteModel(path: string): Promise<ModelArtifactsInfo>;
+
+  /**
+   * Copy a model from one path to another.
+   *
+   * @param oldPath
+   * @param newPath
+   * @returns ModelArtifactsInfo of the copied model (if and only if copying
```
Comment (bileschi) [inline src/io/types.ts]: Same question as above, should these methods return some sort of status object different from the ModelArtifactsInfo themselves?

**Comment by nsthorat** [comment]
Review status: 7 of 9 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/io/indexed_db.ts, line 185 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCdKIX47xoS1sdqT4mG:-LCdKIX47xoS1sdqT4mH:b2k34q5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/209241f9aeab1172babb256bc34ca1a82f02a6d2/src/io/indexed_db.ts#L185)):*
> ```XML
>             });
>             putModelRequest.onsuccess = () => {
>               resolve({modelArtifactsInfo});
> ```

you can one-line this

---

*[src/io/indexed_db.ts, line 406 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCdW4d7E_wUH0cGGOrr:-LCdW4d7E_wUH0cGGOrs:b9syzh4) ([raw file](https://github.com/tensorflow/tfjs-core/blob/209241f9aeab1172babb256bc34ca1a82f02a6d2/src/io/indexed_db.ts#L406)):*
> ```XML
>  * // Delete a model; list models again.
>  * await manager.deleteModel('demo/local-storage-manager/model1');
>  * console.log(await manager.listModels());
> ```

is the idea here that you're going to expand the API?

Why not just have a top-level list method that takes a storage mechanism and lists the model names. Then you can have another "remove" method that takes a the string path like `tf.io.deleteModel('indexeddb://my-model');`

---

*[src/io/io.ts, line 41 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCcRxnA4U_RPucl2sqh:-LCdWQiI93Wf3PxsqV6H:b-nlwc3w) ([raw file](https://github.com/tensorflow/tfjs-core/blob/209241f9aeab1172babb256bc34ca1a82f02a6d2/src/io/io.ts#L41)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

no need to expose these to the API. Instead expose global `listModels/removeModel` methods that take storage type as protocol prefix.
</blockquote></details>

+1, see my comment above

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038)*

**Comment by bileschi** [comment]
Review status: 3 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/io/model_management.ts, line 65 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCeWHv02P8r3e4h8RCt:-LCeWHv02P8r3e4h8RCu:b-8lzmon) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management.ts#L65)):*
> ```XML
>     }
>     assert(scheme.length > 0, 'scheme must not be an empty string.');
>     ModelStoreManagerRegistry.getInstance().managers[scheme] = manager;
> ```

protect against multiple managers attempting to use the same scheme?

---

*[src/io/model_management_test.ts, line 268 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCe_mlC4umuSeWUlQYQ:-LCe_mlC4umuSeWUlQYR:biqi8hd) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management_test.ts#L268)):*
> ```
> Successful moveModel between mediums
> ```
moveModel with identical src/dest medium has a special code path.  Is there a test that this works?  I could have missed it.

---

*[src/io/model_management_test.ts, line 316 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCeZzImBfTe3UhM0XL5:-LCeZzImBfTe3UhM0XL6:b-uasm6g) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management_test.ts#L316)):*
> ```
> to 
> ```
nit: copying *from* invalid URL ...

---

*[src/io/model_management_test.ts, line 367 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCe_HAr0PxkeuPjoliA:-LCe_HAr0PxkeuPjoliB:b-a74s0t) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management_test.ts#L367)):*
> ```
>  done();
> ```
suggestion : test that the source has not been deleted in this case.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038)*

**Comment by bileschi** [comment]
Great!   Some small comments.

---







*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-:-LCe_xp32Lr8BBYyXNFf:b-9gdvoy)*

**Comment by bileschi** [comment]
:lgtm_strong:

---







*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-:-LCea0vt0AOfS6gvYCzZ:b-jbsf8z)*

**Comment by caisq** [comment]
Review status: 3 of 11 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[src/io/model_management.ts, line 65 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCeWHv02P8r3e4h8RCt:-LCf8HkVAKZgXB2DjZer:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management.ts#L65)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

protect against multiple managers attempting to use the same scheme?
</blockquote></details>

Done.

---

*[src/io/model_management_test.ts, line 268 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCe_mlC4umuSeWUlQYQ:-LCf7boTBcDVzmAuKila:bexaix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management_test.ts#L268)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> Successful moveModel between mediums
> ```
moveModel with identical src/dest medium has a special code path.  Is there a test that this works?  I could have missed it.
</blockquote></details>

Moving / copying model within the same medium is covered in the unit tests in local_storage_test.ts and indexed_db_test.ts.

---

*[src/io/model_management_test.ts, line 316 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCeZzImBfTe3UhM0XL5:-LCf7v9S-af1X9iJSJxw:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management_test.ts#L316)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> to 
> ```
nit: copying *from* invalid URL ...
</blockquote></details>

Done.

---

*[src/io/model_management_test.ts, line 367 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-LCe_HAr0PxkeuPjoliA:-LCf7a427LMBsUz_CugK:bcujtxn) ([raw file](https://github.com/tensorflow/tfjs-core/blob/e348870efbb801821fb64b02f00858116ffc248b/src/io/model_management_test.ts#L367)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  done();
> ```
suggestion : test that the source has not been deleted in this case.
</blockquote></details>

Good suggestion! Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038)*

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: 3 of 11 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1038#-:-LCiIxIu7GPQk30LeM5h:b-jbsf8z)*

**Pull request merged**